{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:Error in loading the saved optimizer state. As a result, your model is starting with a freshly initialized optimizer.\n",
      "WARNING:tensorflow:No training configuration found in the save file, so the model was *not* compiled. Compile it manually.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 100/100 [03:59<00:00,  2.39s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes:\n",
      "['VPS13C', 'CCL17', 'RAB3C', 'OXR1', 'RTN1', 'CFAP161', 'WNT10B', 'GPSM2']\n",
      "Average Initial Expression: -0.14392519189277664\n",
      "Best Expression: 0.0003147315396745398\n"
     ]
    }
   ],
   "source": [
    "from re import A, L\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import torch\n",
    "import tensorflow as tf\n",
    "import tensorflow.keras.backend as K\n",
    "from tensorflow.keras import Model\n",
    "from tensorflow.keras.models import load_model\n",
    "import sys\n",
    "sys.path.append('/home/sina/ml/gan/dev/shot0')\n",
    "sys.path.insert(0, '/home/sina/ml/gan/dev/shot0/lib')\n",
    "import argparse\n",
    "from util import *\n",
    "from framepool import *\n",
    "\n",
    "tf.compat.v1.enable_eager_execution()\n",
    "\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests, sys\n",
    "from popen import Auto_popen\n",
    "\n",
    "abs_path = './../mrl_te_optimization/log/Backbone/RL_hard_share/3M/small_repective_filed_strides1113.ini'\n",
    "Configuration = Auto_popen(abs_path)\n",
    "\n",
    "np.random.seed(25)\n",
    "\n",
    "BATCH_SIZE = 64\n",
    "N_GENES = 8\n",
    "LR = 0.0001\n",
    "GPU = '-1'\n",
    "STEPS = 100\n",
    "\n",
    "if GPU == '-1':\n",
    "    device = 'cpu'\n",
    "else:\n",
    "    os.environ['CUDA_VISIBLE_DEVICES'] = GPU\n",
    "    device = 'cuda'\n",
    "    if GPU.includes(','):\n",
    "        device = 'cuda:1'\n",
    "\n",
    "SEQ_BATCH = N_GENES\n",
    "UTR_LEN = 128\n",
    "DIM = 40\n",
    "gpath = './../../models/checkpoint_3000.h5'\n",
    "mrl_path = './../../models/utr_model_combined_residual_new.h5'\n",
    "exp_path = './../../models/humanMedian_trainepoch.11-0.426.h5'\n",
    "tpath = './script/checkpoint/RL_hard_share_MTL/3R/schedule_MTL-model_best_cv1.pth'\n",
    "LR = np.exp(-int(LR))\n",
    "\n",
    "def fetch_seq(start, end, chr, strand):\n",
    "    server = \"https://rest.ensembl.org\"\n",
    "    \n",
    "    ext = \"/sequence/region/human/\" + str(chr) + \":\" + str(start) + \"..\" + str(end) + \":\" + str(strand) +  \"?\"\n",
    "    \n",
    "    r = requests.get(server+ext, headers={ \"Content-Type\" : \"text/plain\"})\n",
    "    \n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "        sys.exit()\n",
    "\n",
    "    return r.text\n",
    "\n",
    "def parse_biomart(path = 'reference_gene_info.txt'):\n",
    "\n",
    "    file = path\n",
    "\n",
    "    fasta_sequences = SeqIO.parse(open(file),'fasta')\n",
    "\n",
    "    genes = []\n",
    "    ustarts = []\n",
    "    uends = []\n",
    "    seqs = []\n",
    "    strands = []\n",
    "    tsss = []\n",
    "    chromosomes = []\n",
    "\n",
    "    counter = 0\n",
    "\n",
    "    for fasta in fasta_sequences:\n",
    "\n",
    "        name, sequence = fasta.id, str(fasta.seq)\n",
    "        \n",
    "        if sequence != \"Sequenceunavailable\":\n",
    "\n",
    "            counter += 1\n",
    "\n",
    "            listed = name.split('|')\n",
    "            # print(len(listed))\n",
    "\n",
    "            if len(listed) == 8:\n",
    "\n",
    "                genes.append(listed[1])\n",
    "                chromosomes.append(listed[2])\n",
    "                \n",
    "                if ';' in listed[3]:\n",
    "                    ustart = listed[3].split(';')[0]\n",
    "                    uend = listed[4].split(';')[0]\n",
    "                else:\n",
    "                    ustart = listed[3]\n",
    "                    uend = listed[4]\n",
    "\n",
    "                strand = int(listed[0])\n",
    "\n",
    "                if strand == -1:\n",
    "                    ustarts.append(ustart)\n",
    "                    uends.append(uend)\n",
    "                else:\n",
    "                    ustarts.append(uend)\n",
    "                    uends.append(ustart)\n",
    "\n",
    "\n",
    "                strands.append(str(strand))\n",
    "                \n",
    "                tsss.append(listed[-1])\n",
    "                \n",
    "                seqs.append(sequence)\n",
    "\n",
    "    df = pd.DataFrame({'utr':seqs,'gene':genes, 'chr': chromosomes,'utr_start':ustarts,'utr_end':uends,'tss':tsss,'strand':strands})\n",
    "    return df\n",
    "\n",
    "\n",
    "def convert_model(model_:Model):\n",
    "    # print(model_.summary())\n",
    "    input_ = tf.keras.layers.Input(shape=( 10500, 4))\n",
    "    input = input_\n",
    "    for i in range(len(model_.layers)-1):\n",
    "\n",
    "        # print(type(model_.layers[i+1]))\n",
    "        \n",
    "        if isinstance(model_.layers[i+1],tf.keras.layers.Concatenate):\n",
    "            paddings = tf.constant([[0,0],[0,6]])\n",
    "            output = tf.pad(input, paddings, 'CONSTANT')\n",
    "            input = output\n",
    "        else:\n",
    "            if not isinstance(model_.layers[i+1],tf.keras.layers.InputLayer):\n",
    "                output = model_.layers[i+1](input)\n",
    "                input = output\n",
    "\n",
    "            if isinstance(model_.layers[i+1],tf.keras.layers.Conv1D):\n",
    "                pass\n",
    "\n",
    "    model = tf.keras.Model(inputs=input_, outputs=output)\n",
    "    model.compile(loss=\"mse\", optimizer=\"adam\")\n",
    "    return model\n",
    "\n",
    "def one_hot(seq):\n",
    "    convert = False\n",
    "    if isinstance(seq, tf.Tensor):\n",
    "        seq = seq.numpy().astype(str)\n",
    "        convert = True\n",
    "\n",
    "    num_seqs = len(seq)\n",
    "    seq_len = len(seq[0])\n",
    "    seqindex = {'A':0, 'C':1, 'G':2, 'T':3, 'a':0, 'c':1, 'g':2, 't':3}\n",
    "    seq_vec = np.zeros((num_seqs,seq_len,4), dtype='bool')\n",
    "    for i in range(num_seqs):\n",
    "        thisseq = seq[i]\n",
    "        for j in range(seq_len):\n",
    "            try:\n",
    "                seq_vec[i,j,seqindex[thisseq[j]]] = 1\n",
    "            except:\n",
    "                pass\n",
    "    \n",
    "    if convert:\n",
    "        seq_vec = tf.convert_to_tensor(seq_vec,dtype=tf.float32)\n",
    "\n",
    "\n",
    "    return seq_vec\n",
    "\n",
    "\n",
    "def read_genes(file_name):\n",
    "    with open(file_name) as f:\n",
    "        lines = f.readlines()\n",
    "\n",
    "    lines = [lines[i].replace('\\n','') for i in range(len(lines))]\n",
    "\n",
    "    return np.array(lines)\n",
    "    \n",
    "def select_dna_single(fname='./../../data/genes.txt',batch_size=64):\n",
    "    refs = read_genes(fname)\n",
    "    indice = random.sample(range(0,refs.shape[0]),1)\n",
    "    refs = refs\n",
    "    return indice[0], refs    \n",
    "\n",
    "def select_dna_batch(fname='./../../data/genes.txt',dna_batch=32):\n",
    "    refs = read_genes(fname)\n",
    "    indices = random.sample(range(0,refs.shape[0]),dna_batch)\n",
    "    refs = refs\n",
    "    gene_names = []\n",
    "    for i in range(len(indices)):\n",
    "        gene_names.append(df.iloc[indices[i]]['gene'])\n",
    "    return indices, refs, gene_names\n",
    "\n",
    "def replace_xpresso_seqs(gens,df:pd.DataFrame,refs,index):\n",
    "    seqs = []\n",
    "\n",
    "    for i in range(len(gens)):\n",
    "        length = len(gens[i])        \n",
    "      \n",
    "        len_ = abs(int(df.iloc[index]['utr_start']) - int(df.iloc[index]['utr_end']))\n",
    "\n",
    "        diff = length - len_\n",
    "        origin_dna = refs[index]            \n",
    "        if len(origin_dna) != 10628:\n",
    "            origin_dna = origin_dna + (10628-len(origin_dna)) * \"A\"\n",
    "                \n",
    "        gen_dna = origin_dna[:7000] + gens[i] + origin_dna[7000+abs(int(df.iloc[index]['utr_start']) - int(df.iloc[index]['utr_end'])):10500 - diff]\n",
    "        if len(gen_dna) != 10500:\n",
    "            gen_dna = gen_dna + (10500-len(gen_dna)) * \"A\"\n",
    "\n",
    "        seqs.append(gen_dna)\n",
    "\n",
    "    seqs = tf.convert_to_tensor(seqs)\n",
    "    return seqs, []\n",
    "\n",
    "def replace_xpresso_seqs_single_v2(gens,df:pd.DataFrame,refs,indices,batch_size=SEQ_BATCH):\n",
    "    seqs = []\n",
    "    # originals = []\n",
    "    ref_len = len(indices)  \n",
    "\n",
    "    for i in range(len(gens)):\n",
    "        length = len(gens[i])        \n",
    "        for j in range(len(indices)):\n",
    "            index = indices[j]            \n",
    "            len_ = abs(int(df.iloc[index]['utr_start']) - int(df.iloc[index]['utr_end']))\n",
    "\n",
    "            diff = length - len_\n",
    "            origin_dna = refs[j]            \n",
    "            if len(origin_dna) != 10628:\n",
    "                origin_dna = origin_dna + (10628-len(origin_dna)) * \"A\"\n",
    "                    \n",
    "            gen_dna = origin_dna[:7000] + gens[i] + origin_dna[7000+abs(int(df.iloc[index]['utr_start']) - int(df.iloc[index]['utr_end'])):10500 - diff]\n",
    "            if len(gen_dna) != 10500:\n",
    "                gen_dna = gen_dna + (10500-len(gen_dna)) * \"A\"\n",
    "\n",
    "            seqs.append(gen_dna)\n",
    "\n",
    "    seqs = tf.convert_to_tensor(seqs)\n",
    "    return seqs, []\n",
    "\n",
    "def select_best(scores, seqs, gc_control=False, GC=-1):\n",
    "    selected_scores = []\n",
    "    selected_seqs = []\n",
    "    for i in range(len(scores[0])):\n",
    "        best = scores[1][i]\n",
    "        best_seq = seqs[1][i]\n",
    "        for j in range(len(scores)-1):\n",
    "            if scores[j+1][i] > best:\n",
    "                if gc_control:\n",
    "                    if get_gc_content(seqs[j][i]) < GC:\n",
    "                        best = scores[j+1][i]\n",
    "                        best_seq = seqs[j+1][i]\n",
    "                else:\n",
    "                    best = scores[j+1][i]\n",
    "                    best_seq = seqs[j+1][i]\n",
    "\n",
    "        selected_scores.append(best)\n",
    "        selected_seqs.append(best_seq)\n",
    "\n",
    "    return selected_seqs, selected_scores\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    df = parse_biomart('./../../data/gene_info.txt')\n",
    "\n",
    "    # Genes retreived from Ensemble\n",
    "    df = df[:8200]\n",
    "\n",
    "    model = tf.keras.models.load_model(exp_path)\n",
    "\n",
    "    model = convert_model(model)\n",
    "\n",
    "    wgan = tf.keras.models.load_model(gpath)\n",
    "\n",
    "    \"\"\"\n",
    "    Data:\n",
    "    \"\"\"\n",
    "\n",
    "    noise = tf.Variable(tf.random.normal(shape=[BATCH_SIZE,40]))\n",
    "\n",
    "    tf.random.set_seed(25)\n",
    "\n",
    "    diffs = []\n",
    "    init_exps = []\n",
    "\n",
    "    opt_exps = []\n",
    "\n",
    "    orig_vals = []\n",
    "\n",
    "    noise = tf.Variable(tf.random.normal(shape=[BATCH_SIZE,40]))\n",
    "    noise_small = tf.random.normal(shape=[BATCH_SIZE,40],stddev=1e-5)\n",
    "\n",
    "    optimizer = tf.keras.optimizers.Adam(learning_rate=LR)\n",
    "\n",
    "    '''\n",
    "    Optimization takes place here.\n",
    "    '''\n",
    "\n",
    "    bind_scores_list = []\n",
    "    bind_scores_means = []\n",
    "    sequences_list = []\n",
    "\n",
    "    means = []\n",
    "    maxes = []\n",
    "\n",
    "    iters_ = []\n",
    "\n",
    "    OPTIMIZE = True\n",
    "\n",
    "    DNA_SEL = False\n",
    "    indices, refs, names = select_dna_batch(dna_batch=SEQ_BATCH)\n",
    "\n",
    "    sequences_init = wgan(noise)\n",
    "\n",
    "    gen_seqs_init = sequences_init.numpy().astype('float')\n",
    "\n",
    "    seqs_gen_init = recover_seq(gen_seqs_init, rev_rna_vocab)\n",
    "\n",
    "    seqs_init, origs = replace_xpresso_seqs_single_v2(seqs_gen_init,df,refs,indices)\n",
    "\n",
    "    seqs_init = one_hot(seqs_init)\n",
    "\n",
    "\n",
    "    pred_init = model(seqs_init) \n",
    "\n",
    "    pred_init = tf.reshape(pred_init,(SEQ_BATCH,-1))\n",
    "\n",
    "    init_t = tf.reduce_mean(pred_init,axis=0)\n",
    "\n",
    "    init_t = init_t.numpy().astype('float')\n",
    "\n",
    "    mrl_model = load_framepool()\n",
    "    te_model = torch.load(tpath,map_location=torch.device(device))['state_dict']      \n",
    "    te_model.train().to(device)\n",
    "\n",
    "\n",
    "\n",
    "    ########### MRL and TE check before optimization ###################\n",
    "\n",
    "    seqs_mrl = tf.convert_to_tensor(np.array([encode_seq_framepool(seq) for seq in seqs_gen_init]),dtype=tf.float32)\n",
    "    seqs_te =  torch.transpose(torch.tensor(np.array(one_hot_all_motif(seqs_gen_init),dtype=np.float32)),2,1).float().to(device)\n",
    "\n",
    "    mrl_preds_init = mrl_model(seqs_mrl).numpy().astype('float')\n",
    "    te_preds_init = te_model.forward(seqs_te).cpu().data.numpy()\n",
    "\n",
    "    ####################################################################\n",
    "\n",
    "    STEPS = STEPS\n",
    "\n",
    "    seqs_collection = []\n",
    "    scores_collection = []\n",
    "    if OPTIMIZE:\n",
    "\n",
    "        \n",
    "        iter_ = 0\n",
    "        for opt_iter in tqdm(range(STEPS)):\n",
    "            \n",
    "            with tf.GradientTape() as gtape:\n",
    "                gtape.watch(noise)\n",
    "                \n",
    "                sequences = wgan(noise)\n",
    "\n",
    "                seqs_gen = recover_seq(sequences, rev_rna_vocab)\n",
    "                seqs_collection.append(seqs_gen)\n",
    "\n",
    "                g1_ = tf.zeros_like(sequences)\n",
    "\n",
    "                scores_collection_temp = []\n",
    "                means_temp = []\n",
    "                maxes_temp = []\n",
    "\n",
    "\n",
    "                for indx in range(len(indices)):\n",
    "\n",
    "                    seqs2, origs = replace_xpresso_seqs(seqs_gen,df,refs,indices[indx])\n",
    "                \n",
    "                    seqs = one_hot(seqs2)\n",
    "                    \n",
    "                    with tf.GradientTape() as ptape:\n",
    "                        ptape.watch(seqs)\n",
    "\n",
    "                        pred =  model(seqs)\n",
    "                        t = tf.reshape(pred,(-1))\n",
    "                        mx = np.amax(t.numpy().astype('float'),axis=0)\n",
    "                        mx = np.max(mx)\n",
    "                        \n",
    "\n",
    "                        scores_collection_temp.append(t.numpy().astype('float'))\n",
    "                        nt = t.numpy().astype('float')\n",
    "                        maxes_temp.append(mx)\n",
    "                        means_temp.append(np.sum(t)/BATCH_SIZE)\n",
    "\n",
    "                    g1 = ptape.gradient(pred,seqs)\n",
    "                    g1 = tf.math.scalar_mul(-1.0, g1)\n",
    "                    g1 = tf.slice(g1,[0,7000,0],[-1,128,-1])\n",
    "\n",
    "                    tmp_g = g1.numpy().astype('float')\n",
    "                    tmp_seqs = seqs_gen\n",
    "\n",
    "                    tmp_lst = np.zeros(shape=(BATCH_SIZE,128,5))\n",
    "                    for i in range(len(tmp_seqs)):\n",
    "                        len_ = len(tmp_seqs[i])\n",
    "                        edited_g = tmp_g[i][:len_,:]\n",
    "                        edited_g = np.pad(edited_g,((0,128-len_),(0,1)),'constant')   \n",
    "                        tmp_lst[i] = edited_g\n",
    "\n",
    "                    g1 = tf.convert_to_tensor(tmp_lst,dtype=tf.float32)\n",
    "\n",
    "                    g1_ = tf.math.add(g1, g1_)\n",
    "\n",
    "                scores_collection.append(np.mean(scores_collection_temp,axis=0)/BATCH_SIZE)\n",
    "                means.append(np.mean(means_temp))\n",
    "                maxes.append(np.max(maxes_temp))\n",
    "\n",
    "                g2 = gtape.gradient(sequences,noise,output_gradients=g1_)\n",
    "\n",
    "\n",
    "            a1 = g2 + noise_small\n",
    "            change = [(a1,noise)]\n",
    "\n",
    "            optimizer.apply_gradients(change)\n",
    "\n",
    "            iters_.append(iter_)\n",
    "            iter_ += 1\n",
    "\n",
    "        sequences_opt = wgan(noise)\n",
    "\n",
    "        gen_seqs_opt = sequences_opt.numpy().astype('float')\n",
    "\n",
    "        seqs_gen_opt = recover_seq(gen_seqs_opt, rev_rna_vocab)\n",
    "\n",
    "        seqs_opt, origs = replace_xpresso_seqs_single_v2(seqs_gen_opt,df,refs,indices)\n",
    "\n",
    "        seqs_opt = one_hot(seqs_opt)\n",
    "\n",
    "        pred_opt = model(seqs_opt)\n",
    "\n",
    "        pred_opt = tf.reshape(pred_opt,(SEQ_BATCH,-1))\n",
    "\n",
    "\n",
    "        t = tf.reduce_mean(pred_opt,axis=0)\n",
    "        opt_t = t.numpy().astype('float')\n",
    "\n",
    "    ########### MRL and TE check after optimization ####################\n",
    "\n",
    "    seqs_mrl = tf.convert_to_tensor(np.array([encode_seq_framepool(seq) for seq in seqs_gen_opt]),dtype=tf.float32)\n",
    "    seqs_te =  torch.transpose(torch.tensor(np.array(one_hot_all_motif(seqs_gen_opt),dtype=np.float32)),2,1).float().to(device)\n",
    "\n",
    "    mrl_preds_opt = mrl_model(seqs_mrl).numpy().astype('float')\n",
    "    te_preds_opt = te_model.forward(seqs_te).cpu().data.numpy()\n",
    "\n",
    "    ####################################################################\n",
    "\n",
    "    best_seqs, best_scores = select_best(scores_collection, seqs_collection)\n",
    "\n",
    "    with open('./outputs/mul_init_exps.txt', 'w') as f:\n",
    "        for item in init_t:\n",
    "            f.write(f'{item}\\n')\n",
    "\n",
    "    with open('./outputs/mul_best_exps.txt', 'w') as f:\n",
    "        for item in best_scores:\n",
    "            f.write(f'{item}\\n')\n",
    "\n",
    "    with open('./outputs/mul_opt_exps.txt', 'w') as f:\n",
    "        for item in opt_t:\n",
    "            f.write(f'{item}\\n')\n",
    "\n",
    "    with open('./outputs/mul_best_seqs.txt', 'w') as f:\n",
    "        for item in best_seqs:\n",
    "            f.write(f'{item}\\n')\n",
    "\n",
    "    with open('./outputs/mul_init_seqs.txt', 'w') as f:\n",
    "        for item in seqs_gen_init:\n",
    "            f.write(f'{item}\\n')\n",
    "\n",
    "\n",
    "\n",
    "    print(\"Genes:\")\n",
    "    print(names)\n",
    "    print(f\"Average Initial Expression: {np.average(init_t)}\")\n",
    "    print(f\"Best Expression: {np.average(best_scores)}\")\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "utrgan",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
