{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/sbarazan/Documents/bilkent/utrgan/bioinformatics advances/code/UTRGAN/src/exp_optimization\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from re import A, L\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import argparse\n",
    "from tqdm import tqdm\n",
    "import torch\n",
    "import tensorflow as tf\n",
    "import tensorflow.keras.backend as K\n",
    "from tensorflow.keras import Model\n",
    "from tensorflow.keras.models import load_model\n",
    "import sys\n",
    "import socket\n",
    "import datetime\n",
    "import random\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests, sys\n",
    "from util import *\n",
    "from framepool import *\n",
    "from popen import Auto_popen\n",
    "\n",
    "tf.compat.v1.enable_eager_execution()\n",
    "\n",
    "DATA = './../../data/utrdb2.csv'\n",
    "BATCH_SIZE = 64\n",
    "GENE = 'IFNG'\n",
    "GC_LIMIT = 1.00\n",
    "LR = 0.0001\n",
    "GPU = '-1'\n",
    "STEPS = 1000\n",
    "\n",
    "if GPU == '-1':\n",
    "    device = 'cpu'\n",
    "else:\n",
    "    os.environ['CUDA_VISIBLE_DEVICES'] = GPU\n",
    "    device = 'cuda'\n",
    "    if GPU.includes(','):\n",
    "        device = 'cuda:1'\n",
    "\n",
    "DIM = 40\n",
    "SEQ_LEN = 128\n",
    "gpath = './../../models/checkpoint_3000.h5'\n",
    "mrl_path = './../../models/utr_model_combined_residual_new.h5'\n",
    "exp_path = './../../models/humanMedian_trainepoch.11-0.426.h5'\n",
    "tpath = './script/checkpoint/RL_hard_share_MTL/3R/schedule_MTL-model_best_cv1.pth'\n",
    "\n",
    "# exp_path = '/home/sina/UTR/models/GM12878_trainepoch.06-0.5062.h5'\n",
    "\n",
    "CELL_LINE = ''\n",
    "# CELL_LINE = 'K562_'\n",
    "# CELL_LINE = 'GM12878_'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:Error in loading the saved optimizer state. As a result, your model is starting with a freshly initialized optimizer.\n",
      "WARNING:tensorflow:No training configuration found in the save file, so the model was *not* compiled. Compile it manually.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/utrgan/lib/python3.10/site-packages/keras/optimizers/optimizer_v2/adam.py:114: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n",
      "  super().__init__(name, **kwargs)\n",
      "100%|██████████| 1000/1000 [04:15<00:00,  3.91it/s]\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: './outputs/gc_init_exps_IFNG.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 492\u001b[0m\n\u001b[1;32m    488\u001b[0m te_preds_opt \u001b[38;5;241m=\u001b[39m te_model\u001b[38;5;241m.\u001b[39mforward(seqs_te)\u001b[38;5;241m.\u001b[39mcpu()\u001b[38;5;241m.\u001b[39mdata\u001b[38;5;241m.\u001b[39mnumpy()\n\u001b[1;32m    490\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m GC_CONTROL:\n\u001b[0;32m--> 492\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m./outputs/\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mCELL_LINE\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43mgc_init_exps_\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43mgene_name\u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m.txt\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mw\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[1;32m    493\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m item \u001b[38;5;129;01min\u001b[39;00m init_t:\n\u001b[1;32m    494\u001b[0m             f\u001b[38;5;241m.\u001b[39mwrite(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mitem\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m/opt/anaconda3/envs/utrgan/lib/python3.10/site-packages/IPython/core/interactiveshell.py:324\u001b[0m, in \u001b[0;36m_modified_open\u001b[0;34m(file, *args, **kwargs)\u001b[0m\n\u001b[1;32m    317\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m file \u001b[38;5;129;01min\u001b[39;00m {\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m2\u001b[39m}:\n\u001b[1;32m    318\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    319\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIPython won\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt let you open fd=\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfile\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m by default \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    320\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mas it is likely to crash IPython. If you know what you are doing, \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    321\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124myou can use builtins\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m open.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    322\u001b[0m     )\n\u001b[0;32m--> 324\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mio_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: './outputs/gc_init_exps_IFNG.txt'"
     ]
    }
   ],
   "source": [
    "def fetch_seq(start, end, chr, strand):\n",
    "    server = \"https://rest.ensembl.org\"\n",
    "    \n",
    "    ext = \"/sequence/region/human/\" + str(chr) + \":\" + str(start) + \"..\" + str(end) + \":\" + str(strand) +  \"?\"\n",
    "    \n",
    "    r = requests.get(server+ext, headers={ \"Content-Type\" : \"text/plain\"})\n",
    "    \n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "        sys.exit()\n",
    "\n",
    "    return r.text\n",
    "\n",
    "def parse_biomart(path = 'martquery_0721120207_840.txt'):\n",
    "\n",
    "    file = path\n",
    "\n",
    "    fasta_sequences = SeqIO.parse(open(file),'fasta')\n",
    "\n",
    "    genes = []\n",
    "    ustarts = []\n",
    "    uends = []\n",
    "    seqs = []\n",
    "    strands = []\n",
    "    tsss = []\n",
    "    chromosomes = []\n",
    "\n",
    "    counter = 0\n",
    "\n",
    "    for fasta in fasta_sequences:\n",
    "\n",
    "        name, sequence = fasta.id, str(fasta.seq)\n",
    "        \n",
    "        if sequence != \"Sequenceunavailable\":\n",
    "\n",
    "            counter += 1\n",
    "\n",
    "            listed = name.split('|')\n",
    "            # print(len(listed))\n",
    "\n",
    "            if len(listed) == 8:\n",
    "\n",
    "                genes.append(listed[1])\n",
    "                chromosomes.append(listed[2])\n",
    "                \n",
    "                if ';' in listed[3]:\n",
    "                    ustart = listed[3].split(';')[0]\n",
    "                    uend = listed[4].split(';')[0]\n",
    "                else:\n",
    "                    ustart = listed[3]\n",
    "                    uend = listed[4]\n",
    "\n",
    "                strand = int(listed[0])\n",
    "\n",
    "                if strand == -1:\n",
    "                    ustarts.append(ustart)\n",
    "                    uends.append(uend)\n",
    "                else:\n",
    "                    ustarts.append(uend)\n",
    "                    uends.append(ustart)\n",
    "\n",
    "\n",
    "                strands.append(str(strand))\n",
    "                \n",
    "                tsss.append(listed[-1])\n",
    "                \n",
    "                seqs.append(sequence)\n",
    "\n",
    "    df = pd.DataFrame({'utr':seqs,'gene':genes, 'chr': chromosomes,'utr_start':ustarts,'utr_end':uends,'tss':tsss,'strand':strands})\n",
    "    return df\n",
    "\n",
    "def convert_model(model_:Model):\n",
    "    input_ = tf.keras.layers.Input(shape=( 10500, 4))\n",
    "    input = input_\n",
    "    for i in range(len(model_.layers)-1):\n",
    "\n",
    "        \n",
    "        if isinstance(model_.layers[i+1],tf.keras.layers.Concatenate):\n",
    "            paddings = tf.constant([[0,0],[0,6]])\n",
    "            output = tf.pad(input, paddings, 'CONSTANT')\n",
    "            input = output\n",
    "        else:\n",
    "            if not isinstance(model_.layers[i+1],tf.keras.layers.InputLayer):\n",
    "                output = model_.layers[i+1](input)\n",
    "                input = output\n",
    "\n",
    "            if isinstance(model_.layers[i+1],tf.keras.layers.Conv1D):\n",
    "                pass\n",
    "\n",
    "    model = tf.keras.Model(inputs=input_, outputs=output)\n",
    "    model.compile(loss=\"mse\", optimizer=\"adam\")\n",
    "    return model\n",
    "\n",
    "def one_hot(seq):\n",
    "    convert = False\n",
    "    if isinstance(seq, tf.Tensor):\n",
    "        seq = seq.numpy().astype(str)\n",
    "        convert = True\n",
    "\n",
    "    num_seqs = len(seq)\n",
    "    seq_len = len(seq[0])\n",
    "    seqindex = {'A':0, 'C':1, 'G':2, 'T':3, 'a':0, 'c':1, 'g':2, 't':3}\n",
    "    seq_vec = np.zeros((num_seqs,seq_len,4), dtype='bool')\n",
    "    for i in range(num_seqs):\n",
    "        thisseq = seq[i]\n",
    "        for j in range(seq_len):\n",
    "            try:\n",
    "                seq_vec[i,j,seqindex[thisseq[j]]] = 1\n",
    "            except:\n",
    "                pass\n",
    "    \n",
    "    if convert:\n",
    "        seq_vec = tf.convert_to_tensor(seq_vec,dtype=tf.float32)\n",
    "\n",
    "\n",
    "    return seq_vec\n",
    "\n",
    "def gen_random_dna(len=10500,size=SEQ_LEN):\n",
    "    list_ = ['A','C','G','T']\n",
    "    dnas = []\n",
    "    for i in range(size):\n",
    "            \n",
    "        list_ = ['A','C','G','T']\n",
    "        mydna = 'AGT'\n",
    "        for i in range(len-3):\n",
    "            char = list_[random.randint(0,3)]\n",
    "            mydna = mydna + char\n",
    "        dnas.append(mydna)\n",
    "\n",
    "\n",
    "    return dnas\n",
    "    \n",
    "def select_dna_single(fname='small_seqs.npy',batch_size=64):\n",
    "    refs = np.load(fname)\n",
    "    indice = random.sample(range(0,refs.shape[0]),1)\n",
    "    refs = refs\n",
    "    return indice[0], refs    \n",
    "\n",
    "def recover_seq(samples, rev_charmap):\n",
    "    \"\"\"Convert samples to strings and save to log directory.\"\"\"\n",
    "    if isinstance(samples,tf.Tensor):\n",
    "        samples = samples.numpy()\n",
    "\n",
    "    char_probs = samples\n",
    "    argmax = np.argmax(char_probs, 2)\n",
    "    seqs = []\n",
    "    for line in argmax:\n",
    "        s = \"\".join(rev_charmap[d] for d in line)\n",
    "        s = s.replace('*','')\n",
    "        seqs.append(s)\n",
    "\n",
    "    seqs = np.array(seqs)\n",
    "    return seqs\n",
    "\n",
    "def replace_seqs_coordinate(original_utr_length, original_sequence, genutrs):\n",
    "    replaced = []\n",
    "\n",
    "    for i in range(len(genutrs)):\n",
    "        utr = genutrs[i]\n",
    "        utr_rep = original_sequence[:7000] + utr + original_sequence[7000+original_utr_length:]\n",
    "        utr_rep = utr_rep[:10500]\n",
    "        if len(utr_rep) < 10500:\n",
    "            utr_rep = utr_rep + (10500 - len(utr_rep)) * 'A'\n",
    "\n",
    "        replaced.append(utr_rep)\n",
    "\n",
    "    return replaced\n",
    "\n",
    "def replace_xpresso_seqs_single(gens,df:pd.DataFrame,refs,index):\n",
    "\n",
    "    if isinstance(gens, tf.Tensor):\n",
    "        gens = gens.numpy().astype('str')\n",
    "\n",
    "    seqs = []\n",
    "    originals = []\n",
    "    len_ = abs(int(df.iloc[index]['utr_start']) - int(df.iloc[index]['utr_end']))\n",
    "\n",
    "    origin_dna = refs[index]\n",
    "    \n",
    "    if len(origin_dna) != 10628:\n",
    "        origin_dna = origin_dna + (10628-len(origin_dna)) * \"A\"\n",
    "    \n",
    "\n",
    "    for i in range(len(gens)):\n",
    "        length = len(gens[i])\n",
    "        diff = length - len_\n",
    "        gen_dna = origin_dna[:7000] + gens[i] + origin_dna[7000+abs(int(df.iloc[index]['utr_start']) - int(df.iloc[index]['utr_end'])):10500 - diff]\n",
    "        \n",
    "        original = origin_dna[:10500]\n",
    "        \n",
    "        if len(gen_dna) < 10500:\n",
    "            gen_dna = gen_dna + (10500 - len(gen_dna)) * \"A\"\n",
    "\n",
    "        seqs.append(gen_dna)\n",
    "\n",
    "\n",
    "    seqs = tf.convert_to_tensor(seqs)\n",
    "\n",
    "    original = tf.convert_to_tensor([original for i in range(BATCH_SIZE)])\n",
    "\n",
    "    return seqs, original\n",
    "\n",
    "rna_vocab = {\"A\":0,\n",
    "             \"C\":1,\n",
    "             \"G\":2,\n",
    "             \"U\":3,\n",
    "             \"*\":4}\n",
    "\n",
    "rev_rna_vocab = {v:k for k,v in rna_vocab.items()}\n",
    "\n",
    "def select_best(scores, seqs, gc_control=False, GC=-1):\n",
    "    t = np.max(scores,axis=1)\n",
    "    # print(scores)\n",
    "    maxinds = np.argmax(scores,axis=0)\n",
    "    selected_scores = []\n",
    "    selected_seqs = []\n",
    "    for i in range(len(maxinds)):\n",
    "        selected_seqs.append(seqs[maxinds[i]][i])\n",
    "        selected_scores.append(scores[maxinds[i]][i])\n",
    "\n",
    "\n",
    "    return selected_seqs, selected_scores\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    model = load_model(exp_path)\n",
    "\n",
    "    model = convert_model(model)\n",
    "\n",
    "    gene_name = GENE\n",
    "\n",
    "    if gene_name == 'TLR6':\n",
    "        \n",
    "        UTRSTARTS = [38856761,38829474,38843639]\n",
    "        UTRENDS = [38856817,38829537,38843791]\n",
    "        TSS = 38856817\n",
    "\n",
    "        STRAND = -1\n",
    "        CHR = 4\n",
    "\n",
    "        UTRSTART = UTRSTARTS[0]\n",
    "        UTREND = UTRENDS[0]\n",
    "\n",
    "        # IFNG\n",
    "        UTRSTART = 38856761\n",
    "        UTREND = 38856817\n",
    "\n",
    "        UTRLENGTH = UTREND - UTRSTART\n",
    "\n",
    "    elif gene_name == 'IFNG':\n",
    "\n",
    "        UTRSTARTS = [68159616]\n",
    "        UTRENDS = [68159740]\n",
    "        TSS = 68159740\n",
    "\n",
    "        STRAND = -1\n",
    "        CHR = 12\n",
    "\n",
    "        UTRSTART = UTRSTARTS[0]\n",
    "        UTREND = UTRENDS[0]\n",
    "\n",
    "        UTRLENGTH = UTREND - UTRSTART\n",
    "\n",
    "    elif gene_name == 'TNF':\n",
    "\n",
    "        # TNF\n",
    "        UTRSTARTS = [31575565]\n",
    "        UTRENDS = [31575741]\n",
    "        TSS = 31575565\n",
    "\n",
    "        STRAND = 1\n",
    "        CHR = 6\n",
    "\n",
    "        UTRSTART = UTRSTARTS[0]\n",
    "        UTREND = UTRENDS[0]\n",
    "\n",
    "        UTRLENGTH = UTREND - UTRSTART\n",
    "\n",
    "    elif gene_name == 'TP53':\n",
    "        \n",
    "        # TNF\n",
    "        UTRSTARTS = [7687377,7676595]\n",
    "        UTRENDS = [7687487,7676622]\n",
    "        TSS = 7687487\n",
    "\n",
    "        STRAND = -1\n",
    "        CHR = 17\n",
    "\n",
    "        UTRSTART = UTRSTARTS[0] \n",
    "        UTREND = UTRENDS[0]\n",
    "\n",
    "        UTRLENGTH = abs(UTREND - UTRSTART)\n",
    "\n",
    "    # original_gene_sequence = fetch_seq(start=TSS-7000,end=TSS+3500 + 128,chr=CHR,strand=STRAND)\n",
    "\n",
    "    with open(f'./genes/{gene_name}.txt','r') as f:\n",
    "        original_gene_sequence = f.readline()\n",
    "\n",
    "\n",
    "    wgan = tf.keras.models.load_model(gpath)\n",
    "\n",
    "    \"\"\"\n",
    "    Data:\n",
    "    \"\"\"\n",
    "\n",
    "    noise = tf.Variable(tf.random.normal(shape=[BATCH_SIZE,DIM]))\n",
    "\n",
    "    tf.random.set_seed(25)\n",
    "\n",
    "    np.random.seed(25)\n",
    "\n",
    "    diffs = []\n",
    "    init_exps = []\n",
    "\n",
    "    opt_exps = []\n",
    "\n",
    "    orig_vals = []\n",
    "\n",
    "\n",
    "    noise = tf.Variable(tf.random.normal(shape=[BATCH_SIZE,DIM]))\n",
    "    # noise = tf.random.normal(shape=[BATCH_SIZE,40])\n",
    "    noise_small = tf.random.normal(shape=[BATCH_SIZE,DIM],stddev=1e-5)\n",
    "\n",
    "    optimizer = tf.keras.optimizers.Adam(learning_rate=1e-1)\n",
    "\n",
    "    '''\n",
    "    Original Gene Expression\n",
    "    '''\n",
    "\n",
    "    seqs_orig = one_hot([original_gene_sequence[:10500]])\n",
    "    pred_orig = model(seqs_orig) \n",
    "\n",
    "\n",
    "    '''\n",
    "    Optimization takes place here.\n",
    "    '''\n",
    "\n",
    "\n",
    "\n",
    "    bind_scores_list = []\n",
    "    bind_scores_means = []\n",
    "    sequences_list = []\n",
    "\n",
    "\n",
    "\n",
    "    iters_ = []\n",
    "\n",
    "    OPTIMIZE = True\n",
    "\n",
    "    DNA_SEL = False\n",
    "\n",
    "    sequences_init = wgan(noise)\n",
    "\n",
    "    gen_seqs_init = sequences_init.numpy().astype('float')\n",
    "\n",
    "    seqs_gen_init = recover_seq(gen_seqs_init, rev_rna_vocab)\n",
    "\n",
    "    seqs_init = replace_seqs_coordinate(UTRLENGTH, original_gene_sequence, seqs_gen_init)\n",
    "\n",
    "\n",
    "    seqs_init = one_hot(seqs_init)\n",
    "\n",
    "    pred_init = model(seqs_init) \n",
    "\n",
    "    t = tf.reshape(pred_init,(-1))\n",
    "\n",
    "    init_t = t.numpy().astype('float')\n",
    "    # sum_init = tf.reduce_max(t).numpy().astype('float')\n",
    "    # sum_init = tf.math.argmax()\n",
    "\n",
    "    mrl_model = load_framepool()\n",
    "    te_model = torch.load(tpath,map_location=torch.device(device))['state_dict']  \n",
    "    te_model.train().to(device)\n",
    "\n",
    "    seqs_mrl = tf.convert_to_tensor(np.array([encode_seq_framepool(seq) for seq in seqs_gen_init]),dtype=tf.float32)\n",
    "    seqs_te =  torch.transpose(torch.tensor(np.array(one_hot_all_motif(seqs_gen_init),dtype=np.float32)),2,1).float().to(device)\n",
    "\n",
    "    mrl_preds_init = mrl_model(seqs_mrl).numpy().astype('float')\n",
    "    te_preds_init = te_model.forward(seqs_te).cpu().data.numpy()\n",
    "\n",
    "    means = []\n",
    "    maxes = []\n",
    "    \n",
    "    STEPS = STEPS\n",
    "\n",
    "    seqs_collection = []\n",
    "    scores_collection = []\n",
    "\n",
    "    GC_CONTROL = False\n",
    "    GC_Limit = 0.65\n",
    "\n",
    "    if GC_LIMIT > 0.:\n",
    "        GC_CONTROL = True\n",
    "\n",
    "    if OPTIMIZE:\n",
    "        \n",
    "        iter_ = 0\n",
    "        for opt_iter in tqdm(range(STEPS)):\n",
    "            \n",
    "            with tf.GradientTape() as gtape:\n",
    "\n",
    "                gtape.watch(noise)\n",
    "                \n",
    "                sequences = wgan(noise)\n",
    "\n",
    "                seqs_gen = recover_seq(sequences, rev_rna_vocab)\n",
    "                seqs_collection.append(seqs_gen)\n",
    "\n",
    "                seqs2 = replace_seqs_coordinate(UTRLENGTH, original_gene_sequence, seqs_gen)\n",
    "            \n",
    "                seqs = one_hot(seqs2)\n",
    "                seqs = tf.convert_to_tensor(seqs,dtype=tf.float32)\n",
    "\n",
    "\n",
    "                with tf.GradientTape() as ptape:\n",
    "\n",
    "                    ptape.watch(seqs)\n",
    "\n",
    "                    pred =  model(seqs)\n",
    "                    t = tf.reshape(pred,(-1))\n",
    "                    scores_collection.append(t.numpy().astype('float'))\n",
    "                    mx = np.amax(t.numpy().astype('float'),axis=0)\n",
    "                    mx = np.max(mx)\n",
    "                    \n",
    "                    sum_ = tf.reduce_sum(t)\n",
    "                    maxes.append(mx)\n",
    "                    means.append(sum_/BATCH_SIZE)\n",
    "                    pred = tf.math.scalar_mul(-1.0, pred)\n",
    "\n",
    "                g1 = ptape.gradient(pred,seqs)\n",
    "\n",
    "                g1 = tf.slice(g1,[0,7000,0],[-1,SEQ_LEN,-1])\n",
    "\n",
    "\n",
    "                tmp_g = g1.numpy().astype('float')\n",
    "                tmp_seqs = seqs_gen\n",
    "\n",
    "                tmp_lst = np.zeros(shape=(BATCH_SIZE,SEQ_LEN,5))\n",
    "                for i in range(len(tmp_seqs)):\n",
    "                    len_ = len(tmp_seqs[i])\n",
    "                    \n",
    "                    edited_g = tmp_g[i][:len_,:]\n",
    "\n",
    "                    edited_g = np.pad(edited_g,((0,SEQ_LEN-len_),(0,1)),'constant')   \n",
    "                    \n",
    "                    tmp_lst[i] = edited_g\n",
    "                    \n",
    "                g1 = tf.convert_to_tensor(tmp_lst,dtype=tf.float32)\n",
    "\n",
    "                g2 = gtape.gradient(sequences,noise,output_gradients=g1)\n",
    "\n",
    "            a1 = g2 + noise_small\n",
    "            change = [(a1,noise)]\n",
    "            \n",
    "            optimizer.apply_gradients(change)\n",
    "\n",
    "            iters_.append(iter_)\n",
    "            iter_ += 1\n",
    "\n",
    "        sequences_opt = wgan(noise)\n",
    "\n",
    "        gen_seqs_opt = sequences_opt.numpy().astype('float')\n",
    "\n",
    "        seqs_gen_opt = recover_seq(gen_seqs_opt, rev_rna_vocab)\n",
    "\n",
    "        seqs_opt= replace_seqs_coordinate(UTRLENGTH, original_gene_sequence, seqs_gen_opt)\n",
    "\n",
    "        seqs_opt = one_hot(seqs_opt)\n",
    "\n",
    "        pred_opt = model(seqs_opt)\n",
    "\n",
    "        t = tf.reshape(pred_opt,(-1))\n",
    "        opt_t = t.numpy().astype('float')\n",
    "\n",
    "        # print(seqs_collection)\n",
    "        # print(scores_collection)\n",
    "\n",
    "        if GC_CONTROL:\n",
    "            best_seqs, best_scores = select_best(scores_collection, seqs_collection, True, GC_Limit)\n",
    "        else:\n",
    "            best_seqs, best_scores = select_best(scores_collection, seqs_collection)\n",
    "\n",
    "\n",
    "        seqs_mrl = tf.convert_to_tensor(np.array([encode_seq_framepool(seq) for seq in seqs_gen_opt]),dtype=tf.float32)\n",
    "        seqs_te =  torch.transpose(torch.tensor(np.array(one_hot_all_motif(seqs_gen_opt),dtype=np.float32)),2,1).float().to(device)\n",
    "\n",
    "        mrl_preds_opt = mrl_model(seqs_mrl).numpy().astype('float')\n",
    "        te_preds_opt = te_model.forward(seqs_te).cpu().data.numpy()\n",
    "\n",
    "        if GC_CONTROL:\n",
    "\n",
    "            with open(f'./outputs/{CELL_LINE}gc_init_exps_'+gene_name+'.txt', 'w') as f:\n",
    "                for item in init_t:\n",
    "                    f.write(f'{item}\\n')\n",
    "\n",
    "            with open(f'./outputs/{CELL_LINE}gc_opt_exps_'+gene_name+'.txt', 'w') as f:\n",
    "                for item in best_scores:\n",
    "                    f.write(f'{item}\\n')\n",
    "\n",
    "            with open(f'./outputs/{CELL_LINE}gc_best_seqs_'+gene_name+'.txt', 'w') as f:\n",
    "                for item in best_seqs:\n",
    "                    f.write(f'{item}\\n')\n",
    "\n",
    "            with open(f'./outputs/{CELL_LINE}gc_init_seqs_'+gene_name+'.txt', 'w') as f:\n",
    "                for item in seqs_gen_init:\n",
    "                    f.write(f'{item}\\n')\n",
    "\n",
    "        else:\n",
    "            with open(f'./outputs/{CELL_LINE}init_exps_{gene_name}.txt', 'w') as f:\n",
    "                for item in init_t:\n",
    "                    f.write(f'{item}\\n')\n",
    "\n",
    "            with open(f'./outputs/{CELL_LINE}opt_exps_{gene_name}.txt', 'w') as f:\n",
    "                for item in best_scores:\n",
    "                    f.write(f'{item}\\n')\n",
    "\n",
    "            with open(f'./outputs/{CELL_LINE}best_seqs_{gene_name}.txt', 'w') as f:\n",
    "                for item in best_seqs:\n",
    "                    f.write(f'{item}\\n')\n",
    "\n",
    "            with open(f'./outputs/{CELL_LINE}init_seqs_{gene_name}.txt', 'w') as f:\n",
    "                for item in seqs_gen_init:\n",
    "                    f.write(f'{item}\\n')\n",
    "\n",
    "\n",
    "        print(f\"Average Initial Expression: {np.average(init_t)}\")\n",
    "        print(f\"Best Expression: {np.average(best_scores)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene ID: ENSG00000111537\n",
      "Gene Start: 68154768\n",
      "Gene End: 68159740\n",
      "Strand: -1\n",
      "TSS: 68159740\n",
      "Gene Sequence (first 100 bases): ACATTGTTCTGATCATCTGAAGATCAGCTATTAGAAGAGAAAGATCAGTTAAGTCCTTTGGACCTGATCAGCTTGATACAAGAACTACTGATTTCAACTT ...\n",
      "No UTR information found in the gene region.\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "def get_gene_genomic_info(gene_name, species=\"homo_sapiens\"):\n",
    "    \"\"\"\n",
    "    Given a gene symbol, retrieves gene-level data including:\n",
    "      - Ensembl gene ID,\n",
    "      - Gene start and end coordinates,\n",
    "      - Strand and computed TSS,\n",
    "      - Genomic sequence of the gene,\n",
    "      - UTR segments overlapping the gene region (genomic coordinates).\n",
    "    \"\"\"\n",
    "    base_url = \"https://rest.ensembl.org\"\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "\n",
    "    # Step 1: Get gene ID using the gene symbol\n",
    "    xref_url = f\"{base_url}/xrefs/symbol/{species}/{gene_name}?object_type=gene\"\n",
    "    response = requests.get(xref_url, headers=headers)\n",
    "    if not response.ok:\n",
    "        raise Exception(f\"Error fetching gene ID for symbol {gene_name}: {response.status_code}\")\n",
    "    xref_data = response.json()\n",
    "    if not xref_data:\n",
    "        raise Exception(f\"No gene found for symbol '{gene_name}'.\")\n",
    "    gene_id = xref_data[0]['id']\n",
    "\n",
    "    # Step 2: Retrieve gene-level details (without transcript expansion)\n",
    "    lookup_url = f\"{base_url}/lookup/id/{gene_id}?expand=0\"\n",
    "    response = requests.get(lookup_url, headers=headers)\n",
    "    if not response.ok:\n",
    "        raise Exception(f\"Error fetching details for gene {gene_id}: {response.status_code}\")\n",
    "    gene_data = response.json()\n",
    "\n",
    "    gene_start = gene_data.get(\"start\")\n",
    "    gene_end = gene_data.get(\"end\")\n",
    "    strand = gene_data.get(\"strand\")\n",
    "    seq_region_name = gene_data.get(\"seq_region_name\")\n",
    "    # Determine TSS: gene start if on the plus strand, gene end if on the minus strand\n",
    "    tss = gene_start if strand == 1 else gene_end\n",
    "\n",
    "    # Step 3: Retrieve the gene's genomic sequence\n",
    "    seq_url = f\"{base_url}/sequence/id/{gene_id}?type=genomic\"\n",
    "    response = requests.get(seq_url, headers=headers)\n",
    "    gene_sequence = response.json().get(\"seq\") if response.ok else None\n",
    "\n",
    "    # Step 4: Retrieve UTR features from the gene region using the overlap/region endpoint.\n",
    "    # Note: This returns UTR features from all transcripts overlapping the gene region.\n",
    "    region = f\"{seq_region_name}:{gene_start}-{gene_end}\"\n",
    "    utr_url = f\"{base_url}/overlap/region/{species}/{region}?feature=utr\"\n",
    "    response = requests.get(utr_url, headers=headers)\n",
    "    utr_info = response.json() if response.ok else []\n",
    "    \n",
    "    result = {\n",
    "        \"gene_id\": gene_id,\n",
    "        \"gene_name\": gene_name,\n",
    "        \"gene_start\": gene_start,\n",
    "        \"gene_end\": gene_end,\n",
    "        \"strand\": strand,\n",
    "        \"TSS\": tss,\n",
    "        \"gene_sequence\": gene_sequence,\n",
    "        \"UTRs\": utr_info\n",
    "    }\n",
    "\n",
    "    return result\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    gene_name = input(\"Enter gene name: \").strip()\n",
    "    try:\n",
    "        info = get_gene_genomic_info(gene_name)\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "    else:\n",
    "        print(f\"Gene ID: {info['gene_id']}\")\n",
    "        print(f\"Gene Start: {info['gene_start']}\")\n",
    "        print(f\"Gene End: {info['gene_end']}\")\n",
    "        print(f\"Strand: {info['strand']}\")\n",
    "        print(f\"TSS: {info['TSS']}\")\n",
    "        if info['gene_sequence']:\n",
    "            print(f\"Gene Sequence (first 100 bases): {info['gene_sequence'][:100]} ...\")\n",
    "        else:\n",
    "            print(\"Gene sequence not available.\")\n",
    "        \n",
    "        if info['UTRs']:\n",
    "            print(\"UTR segments (genomic coordinates):\")\n",
    "            for utr in info['UTRs']:\n",
    "                print(f\"  - Chromosome: {utr.get('seq_region_name')}, Start: {utr.get('start')}, End: {utr.get('end')}\")\n",
    "        else:\n",
    "            print(\"No UTR information found in the gene region.\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene: BRCA1 (Ensembl ID: ENSG00000012048)\n",
      "\n",
      "Genome Coordinates:\n",
      "Chromosome: 17\n",
      "Start: 43044295\n",
      "End: 43170245\n",
      "Strand: -\n",
      "\n",
      "Transcription Start Site (TSS):\n",
      "Chromosome: 17\n",
      "Position: 43170245\n",
      "\n",
      "Transcripts:\n",
      "\n",
      "Transcript 1: ENST00000497488\n",
      "Start: 43044295\n",
      "End: 43125300\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125300\n",
      "  5' UTR: 43094643-43094860\n",
      "  3' UTR: 43044295-43045677\n",
      "\n",
      "Transcript 2: ENST00000489037\n",
      "Start: 43044295\n",
      "End: 43125321\n",
      "UTRs:\n",
      "  5' UTR: 43125182-43125321\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044295-43045677\n",
      "\n",
      "Transcript 3: ENST00000478531\n",
      "Start: 43044295\n",
      "End: 43125359\n",
      "UTRs:\n",
      "  5' UTR: 43125277-43125359\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044295-43045677\n",
      "\n",
      "Transcript 4: ENST00000357654\n",
      "Start: 43044295\n",
      "End: 43125364\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125364\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044295-43045677\n",
      "\n",
      "Transcript 5: ENST00000473961\n",
      "Start: 43044295\n",
      "End: 43125364\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125364\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044295-43045677\n",
      "\n",
      "Transcript 6: ENST00000477152\n",
      "Start: 43044295\n",
      "End: 43125364\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125364\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044295-43045677\n",
      "\n",
      "Transcript 7: ENST00000352993\n",
      "Start: 43044295\n",
      "End: 43125370\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125370\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044295-43045677\n",
      "\n",
      "Transcript 8: ENST00000493919\n",
      "Start: 43044295\n",
      "End: 43125402\n",
      "UTRs:\n",
      "  5' UTR: 43125277-43125402\n",
      "  5' UTR: 43124017-43124115\n",
      "  5' UTR: 43106527-43106533\n",
      "  3' UTR: 43044295-43045677\n",
      "\n",
      "Transcript 9: ENST00000494123\n",
      "Start: 43044295\n",
      "End: 43125450\n",
      "UTRs:\n",
      "  5' UTR: 43125277-43125450\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044295-43045677\n",
      "\n",
      "Transcript 10: ENST00000471181\n",
      "Start: 43044295\n",
      "End: 43125483\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125483\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044295-43045677\n",
      "\n",
      "Transcript 11: ENST00000652672\n",
      "Start: 43044295\n",
      "End: 43125483\n",
      "UTRs:\n",
      "  5' UTR: 43125277-43125483\n",
      "  5' UTR: 43124017-43124115\n",
      "  5' UTR: 43121558-43121676\n",
      "  5' UTR: 43115726-43115779\n",
      "  5' UTR: 43106527-43106533\n",
      "  3' UTR: 43044295-43045677\n",
      "\n",
      "Transcript 12: ENST00000634433\n",
      "Start: 43044295\n",
      "End: 43170245\n",
      "UTRs:\n",
      "  5' UTR: 43170126-43170245\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044295-43045677\n",
      "\n",
      "Transcript 13: ENST00000476777\n",
      "Start: 43044298\n",
      "End: 43125360\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125360\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044298-43045677\n",
      "\n",
      "Transcript 14: ENST00000700081\n",
      "Start: 43044302\n",
      "End: 43050409\n",
      "\n",
      "Transcript 15: ENST00000470026\n",
      "Start: 43044304\n",
      "End: 43125343\n",
      "UTRs:\n",
      "  5' UTR: 43125182-43125343\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044304-43045677\n",
      "\n",
      "Transcript 16: ENST00000713676\n",
      "Start: 43044304\n",
      "End: 43125343\n",
      "UTRs:\n",
      "  5' UTR: 43125277-43125343\n",
      "  5' UTR: 43124097-43124102\n",
      "  3' UTR: 43044304-43045677\n",
      "\n",
      "Transcript 17: ENST00000618469\n",
      "Start: 43044304\n",
      "End: 43125364\n",
      "UTRs:\n",
      "  5' UTR: 43124737-43125364\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044304-43045677\n",
      "\n",
      "Transcript 18: ENST00000461574\n",
      "Start: 43044304\n",
      "End: 43125370\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125370\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044304-43045677\n",
      "\n",
      "Transcript 19: ENST00000644555\n",
      "Start: 43044304\n",
      "End: 43125370\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125370\n",
      "  5' UTR: 43124017-43124115\n",
      "  5' UTR: 43119121-43119265\n",
      "  5' UTR: 43115726-43115779\n",
      "  5' UTR: 43106527-43106533\n",
      "  3' UTR: 43044304-43045677\n",
      "\n",
      "Transcript 20: ENST00000468300\n",
      "Start: 43044805\n",
      "End: 43125451\n",
      "UTRs:\n",
      "  5' UTR: 43125277-43125451\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43044805-43045783\n",
      "\n",
      "Transcript 21: ENST00000700082\n",
      "Start: 43044981\n",
      "End: 43048473\n",
      "\n",
      "Transcript 22: ENST00000644379\n",
      "Start: 43045086\n",
      "End: 43125364\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125364\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43045086-43045677\n",
      "\n",
      "Transcript 23: ENST00000484087\n",
      "Start: 43045434\n",
      "End: 43125343\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125343\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43045434-43045677\n",
      "\n",
      "Transcript 24: ENST00000586385\n",
      "Start: 43045563\n",
      "End: 43125329\n",
      "UTRs:\n",
      "  5' UTR: 43125186-43125329\n",
      "  3' UTR: 43045563-43045677\n",
      "\n",
      "Transcript 25: ENST00000591534\n",
      "Start: 43045563\n",
      "End: 43125329\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125329\n",
      "  5' UTR: 43074479-43074521\n",
      "  3' UTR: 43045563-43045677\n",
      "\n",
      "Transcript 26: ENST00000591849\n",
      "Start: 43045563\n",
      "End: 43125329\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125329\n",
      "  5' UTR: 43050093-43050190\n",
      "  3' UTR: 43045563-43045677\n",
      "\n",
      "Transcript 27: ENST00000493795\n",
      "Start: 43045629\n",
      "End: 43125402\n",
      "UTRs:\n",
      "  5' UTR: 43125277-43125402\n",
      "  5' UTR: 43124017-43124115\n",
      "  5' UTR: 43106527-43106533\n",
      "  3' UTR: 43045629-43045677\n",
      "\n",
      "Transcript 28: ENST00000461221\n",
      "Start: 43045678\n",
      "End: 43125288\n",
      "UTRs:\n",
      "  5' UTR: 43125182-43125288\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43104868-43104954\n",
      "  3' UTR: 43104122-43104261\n",
      "  3' UTR: 43099775-43099877\n",
      "  3' UTR: 43097244-43097289\n",
      "  3' UTR: 43095846-43095922\n",
      "  3' UTR: 43091435-43094860\n",
      "  3' UTR: 43090944-43091032\n",
      "  3' UTR: 43082404-43082575\n",
      "  3' UTR: 43076488-43076614\n",
      "  3' UTR: 43074331-43074521\n",
      "  3' UTR: 43070928-43071238\n",
      "  3' UTR: 43067608-43067695\n",
      "  3' UTR: 43063874-43063951\n",
      "  3' UTR: 43063333-43063373\n",
      "  3' UTR: 43057052-43057135\n",
      "  3' UTR: 43051063-43051117\n",
      "  3' UTR: 43049121-43049194\n",
      "  3' UTR: 43047643-43047703\n",
      "  3' UTR: 43045678-43045802\n",
      "\n",
      "Transcript 29: ENST00000491747\n",
      "Start: 43045678\n",
      "End: 43125356\n",
      "UTRs:\n",
      "  5' UTR: 43125277-43125356\n",
      "  5' UTR: 43124097-43124115\n",
      "\n",
      "Transcript 30: ENST00000472490\n",
      "Start: 43067274\n",
      "End: 43071066\n",
      "\n",
      "Transcript 31: ENST00000700182\n",
      "Start: 43070344\n",
      "End: 43125343\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125343\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43070344-43070885\n",
      "\n",
      "Transcript 32: ENST00000621897\n",
      "Start: 43076488\n",
      "End: 43082760\n",
      "\n",
      "Transcript 33: ENST00000354071\n",
      "Start: 43091098\n",
      "End: 43125315\n",
      "\n",
      "Transcript 34: ENST00000700183\n",
      "Start: 43094033\n",
      "End: 43125343\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125343\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43110465-43110534\n",
      "  3' UTR: 43106478-43106533\n",
      "  3' UTR: 43104868-43104956\n",
      "  3' UTR: 43104122-43104261\n",
      "  3' UTR: 43099775-43099880\n",
      "  3' UTR: 43097493-43097586\n",
      "  3' UTR: 43097244-43097289\n",
      "  3' UTR: 43095846-43095922\n",
      "  3' UTR: 43094033-43094860\n",
      "\n",
      "Transcript 35: ENST00000492859\n",
      "Start: 43094112\n",
      "End: 43125300\n",
      "UTRs:\n",
      "  5' UTR: 43125271-43125300\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43110465-43110534\n",
      "  3' UTR: 43106456-43106533\n",
      "  3' UTR: 43104868-43104956\n",
      "  3' UTR: 43104122-43104261\n",
      "  3' UTR: 43099775-43099880\n",
      "  3' UTR: 43097244-43097289\n",
      "  3' UTR: 43095846-43095922\n",
      "  3' UTR: 43094112-43094860\n",
      "\n",
      "Transcript 36: ENST00000642945\n",
      "Start: 43094482\n",
      "End: 43125343\n",
      "UTRs:\n",
      "  5' UTR: 43125277-43125343\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43112487-43112494\n",
      "  3' UTR: 43106456-43106533\n",
      "  3' UTR: 43104868-43104956\n",
      "  3' UTR: 43104122-43104261\n",
      "  3' UTR: 43099775-43099880\n",
      "  3' UTR: 43097244-43097289\n",
      "  3' UTR: 43095846-43095922\n",
      "  3' UTR: 43094482-43094860\n",
      "\n",
      "Transcript 37: ENST00000700184\n",
      "Start: 43099643\n",
      "End: 43124339\n",
      "\n",
      "Transcript 38: ENST00000461798\n",
      "Start: 43099831\n",
      "End: 43125370\n",
      "UTRs:\n",
      "  5' UTR: 43125277-43125370\n",
      "  5' UTR: 43124097-43124115\n",
      "  3' UTR: 43104868-43104954\n",
      "  3' UTR: 43104122-43104261\n",
      "  3' UTR: 43099831-43099880\n",
      "\n",
      "Transcript 39: ENST00000700083\n",
      "Start: 43104225\n",
      "End: 43106139\n",
      "\n",
      "Transcript 40: ENST00000700185\n",
      "Start: 43114391\n",
      "End: 43125370\n",
      "\n",
      "Transcript 41: ENST00000700186\n",
      "Start: 43122343\n",
      "End: 43125370\n",
      "\n",
      "Sequence (first 100 bp):\n",
      "AAAGCGTGGGAATTACAGATAAATTAAAACTGTGGAACCCCTTTCCTCGGCTGCCGCCAAGGTGTTCGGTCCTTCCGAGGAAGCTAAGGCCGCGTTGGGG...\n",
      "Total sequence length: 125951 bp\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "import time\n",
    "\n",
    "class GeneInfoRetriever:\n",
    "    def __init__(self):\n",
    "        self.base_url = \"https://rest.ensembl.org\"\n",
    "        self.headers = {\"Content-Type\": \"application/json\"}\n",
    "        self.sleep_time = 0.5  # To respect Ensembl API rate limits\n",
    "\n",
    "    def get_gene_id(self, gene_symbol, species=\"human\"):\n",
    "        \"\"\"Retrieve the Ensembl gene ID for a gene symbol.\"\"\"\n",
    "        species_map = {\"human\": \"homo_sapiens\", \"mouse\": \"mus_musculus\"}\n",
    "        species_name = species_map.get(species.lower(), species)\n",
    "        \n",
    "        endpoint = f\"/lookup/symbol/{species_name}/{gene_symbol}\"\n",
    "        response = self._make_request(endpoint)\n",
    "        \n",
    "        if response:\n",
    "            return response.get(\"id\")\n",
    "        return None\n",
    "\n",
    "    def get_gene_sequence(self, gene_id):\n",
    "        \"\"\"Retrieve the gene sequence for a gene ID.\"\"\"\n",
    "        endpoint = f\"/sequence/id/{gene_id}?type=genomic\"\n",
    "        response = self._make_request(endpoint)\n",
    "        \n",
    "        if response:\n",
    "            return response.get(\"seq\")\n",
    "        return None\n",
    "\n",
    "    def get_gene_coordinates(self, gene_id):\n",
    "        \"\"\"Retrieve the genomic coordinates for a gene ID.\"\"\"\n",
    "        endpoint = f\"/lookup/id/{gene_id}?expand=1\"\n",
    "        response = self._make_request(endpoint)\n",
    "        \n",
    "        if response:\n",
    "            return {\n",
    "                \"chromosome\": response.get(\"seq_region_name\"),\n",
    "                \"start\": response.get(\"start\"),\n",
    "                \"end\": response.get(\"end\"),\n",
    "                \"strand\": response.get(\"strand\")\n",
    "            }\n",
    "        return None\n",
    "\n",
    "    def get_tss_location(self, gene_id):\n",
    "        \"\"\"\n",
    "        Retrieve the transcription start site (TSS) location.\n",
    "        The TSS is typically at the start of the gene on the positive strand,\n",
    "        or at the end of the gene on the negative strand.\n",
    "        \"\"\"\n",
    "        coords = self.get_gene_coordinates(gene_id)\n",
    "        \n",
    "        if coords:\n",
    "            # For genes on the positive strand, TSS is at the start\n",
    "            # For genes on the negative strand, TSS is at the end\n",
    "            tss = coords[\"start\"] if coords[\"strand\"] == 1 else coords[\"end\"]\n",
    "            return {\n",
    "                \"chromosome\": coords[\"chromosome\"],\n",
    "                \"position\": tss,\n",
    "                \"strand\": coords[\"strand\"]\n",
    "            }\n",
    "        return None\n",
    "\n",
    "    def get_transcript_info(self, gene_id):\n",
    "        \"\"\"Get information about all transcripts for a gene, including UTRs.\"\"\"\n",
    "        # This endpoint needs to be expanded to include UTRs\n",
    "        endpoint = f\"/lookup/id/{gene_id}?expand=1&utr=1\"\n",
    "        response = self._make_request(endpoint)\n",
    "        \n",
    "        transcripts = []\n",
    "        if response and \"Transcript\" in response:\n",
    "            for transcript in response[\"Transcript\"]:\n",
    "                transcript_info = {\n",
    "                    \"transcript_id\": transcript.get(\"id\"),\n",
    "                    \"start\": transcript.get(\"start\"),\n",
    "                    \"end\": transcript.get(\"end\"),\n",
    "                    \"strand\": transcript.get(\"strand\"),\n",
    "                    \"UTRs\": []\n",
    "                }\n",
    "                \n",
    "                # Extract UTR information based on actual Ensembl API response structure\n",
    "                if \"UTR\" in transcript:\n",
    "                    for utr in transcript[\"UTR\"]:\n",
    "                        utr_type = \"5' UTR\" if utr.get(\"object_type\") == \"five_prime_UTR\" else \"3' UTR\"\n",
    "                        transcript_info[\"UTRs\"].append({\n",
    "                            \"start\": utr.get(\"start\"),\n",
    "                            \"end\": utr.get(\"end\"),\n",
    "                            \"type\": utr_type\n",
    "                        })\n",
    "                \n",
    "                transcripts.append(transcript_info)\n",
    "                \n",
    "        return transcripts\n",
    "\n",
    "    def _make_request(self, endpoint):\n",
    "        \"\"\"Make a request to the Ensembl REST API.\"\"\"\n",
    "        url = self.base_url + endpoint\n",
    "        \n",
    "        try:\n",
    "            response = requests.get(url, headers=self.headers)\n",
    "            time.sleep(self.sleep_time)  # Be nice to the API\n",
    "            \n",
    "            if response.status_code == 200:\n",
    "                return response.json()\n",
    "            else:\n",
    "                print(f\"Error: {response.status_code} - {response.text}\")\n",
    "                return None\n",
    "        except Exception as e:\n",
    "            print(f\"Request error: {e}\")\n",
    "            return None\n",
    "\n",
    "    def get_all_gene_info(self, gene_symbol, species=\"human\"):\n",
    "        \"\"\"Retrieve all information about a gene.\"\"\"\n",
    "        # Get the Ensembl gene ID\n",
    "        gene_id = self.get_gene_id(gene_symbol, species)\n",
    "        if not gene_id:\n",
    "            return {\"error\": f\"Gene {gene_symbol} not found\"}\n",
    "        \n",
    "        # Get the gene sequence\n",
    "        sequence = self.get_gene_sequence(gene_id)\n",
    "        \n",
    "        # Get the gene coordinates\n",
    "        coordinates = self.get_gene_coordinates(gene_id)\n",
    "        \n",
    "        # Get the TSS location\n",
    "        tss = self.get_tss_location(gene_id)\n",
    "        \n",
    "        # Get transcript information (including UTRs)\n",
    "        transcripts = self.get_transcript_info(gene_id)\n",
    "        \n",
    "        return {\n",
    "            \"gene_symbol\": gene_symbol,\n",
    "            \"gene_id\": gene_id,\n",
    "            \"sequence\": sequence,\n",
    "            \"coordinates\": coordinates,\n",
    "            \"tss\": tss,\n",
    "            \"transcripts\": transcripts\n",
    "        }\n",
    "\n",
    "def main():\n",
    "    # Predefined variables - change these values as needed\n",
    "    gene_symbol = \"BRCA1\"  # Example gene\n",
    "    species = \"human\"      # Default species\n",
    "    \n",
    "    retriever = GeneInfoRetriever()\n",
    "    result = retriever.get_all_gene_info(gene_symbol, species)\n",
    "    \n",
    "    # Print results\n",
    "    if \"error\" in result:\n",
    "        print(result[\"error\"])\n",
    "    else:\n",
    "        print(f\"Gene: {result['gene_symbol']} (Ensembl ID: {result['gene_id']})\")\n",
    "        print(\"\\nGenome Coordinates:\")\n",
    "        print(f\"Chromosome: {result['coordinates']['chromosome']}\")\n",
    "        print(f\"Start: {result['coordinates']['start']}\")\n",
    "        print(f\"End: {result['coordinates']['end']}\")\n",
    "        print(f\"Strand: {'+' if result['coordinates']['strand'] == 1 else '-'}\")\n",
    "        \n",
    "        print(\"\\nTranscription Start Site (TSS):\")\n",
    "        print(f\"Chromosome: {result['tss']['chromosome']}\")\n",
    "        print(f\"Position: {result['tss']['position']}\")\n",
    "        \n",
    "        print(\"\\nTranscripts:\")\n",
    "        for i, transcript in enumerate(result['transcripts'], 1):\n",
    "            print(f\"\\nTranscript {i}: {transcript['transcript_id']}\")\n",
    "            print(f\"Start: {transcript['start']}\")\n",
    "            print(f\"End: {transcript['end']}\")\n",
    "            \n",
    "            if transcript['UTRs']:\n",
    "                print(\"UTRs:\")\n",
    "                for utr in transcript['UTRs']:\n",
    "                    print(f\"  {utr['type']}: {utr['start']}-{utr['end']}\")\n",
    "        \n",
    "        print(\"\\nSequence (first 100 bp):\")\n",
    "        if result['sequence']:\n",
    "            print(result['sequence'][:100] + \"...\")\n",
    "            print(f\"Total sequence length: {len(result['sequence'])} bp\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "utrgan",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
